projects/asn1-proto/asn1_proto_converter.cc

```diff
@@ -0,0 +1,134 @@
+#include "asn1_proto_converter.h"
+
+namespace asn1_proto {
+
+uint8_t ASN1ProtoConverter::GetNumBytes(size_t num) {
+  uint8_t shift = sizeof(size_t);
+  while (shift != 0) {
+    if (((num >> (shift * 8)) & 0xFF) != 0) {
+      return shift;
+    }
+    shift--;
+  }
+  return 0;
+}
+
+void ASN1ProtoConverter::AppendBytes(size_t num, size_t pos) {
+  uint8_t len_num_bytes = GetNumBytes(num);
+  std::vector<uint8_t> len_vec;
+  for (uint8_t i = len_num_bytes; i != 0; i--) {
+    len_vec.push_back((num >> (i * 7)) & 0xFF);
+  }
+  len_vec.push_back((num & 0xFF));
+  encoder_.insert(encoder_.begin() + pos, len_vec.begin(), len_vec.end());
+}
+
+size_t ASN1ProtoConverter::LongForm(const size_t assigned_len,
+                                    const size_t len_pos) {
+  uint8_t len_bytes = GetNumBytes(assigned_len);
+  if (assigned_len > 127) {
+    uint8_t longForm = (1 << 7);
+    longForm += len_bytes;
+    AppendBytes(longForm, len_pos);
+    return len_bytes;
+  }
+  return 0;
+}
+
+size_t ASN1ProtoConverter::ParseLength(const Length &len,
+                                       const size_t actual_len,
+                                       const size_t len_pos) {
+  size_t assigned_len = 0;
+  if (len.has_length_override()) {
+    sscanf(len.length_override().c_str(), "%zu", &assigned_len);
+  } else if (len.has_indefinite_form() && len.indefinite_form()) {
+    assigned_len = 0x80;
+    // value is placed before length
+    // so the pdu's value is already in encoder
+    // we push 0x00 0x00 (End-of-Content) for indefinite form
+    // which is considered a zero-length object so we
+    // need not add anything to the assigned_len
+    AppendBytes(0x00, encoder_.size());
+    AppendBytes(0x00, encoder_.size());
+  } else {
+    assigned_len = actual_len;
+  }
+  AppendBytes(assigned_len, len_pos);
+  size_t long_len_bytes = LongForm(assigned_len, len_pos);
+  return GetNumBytes(assigned_len) + long_len_bytes;
+}
+
+size_t ASN1ProtoConverter::ParseValue(const Value &val) {
+  size_t len;
+  if (val.val_array().size() > 0) {
+    for (const auto &val_ele : val.val_array()) {
+      if (val_ele.has_pdu()) {
+        len += ParsePDU(val_ele.pdu());
+      } else if (val_ele.has_val_bits()) {
+        len = val_ele.val_bits().size();
+        encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
+                        val_ele.val_bits().end());
+      } else {
+        len = 0;
+      }
+    }
+  }
+  return len;
+}
+
+size_t ASN1ProtoConverter::HighTagForm(const uint8_t cls, const uint8_t enc,
```

Comment:
```suggestion
uint64_t ASN1ProtoConverter::HighTagForm(const uint8_t cls, const uint8_t enc,
```

We should distinguish when you're using a 64-bit integer explicitly, and when you're representing a size of bytes / offset. This probably will have cascading effects, as I think you've got a few overloaded functions here related.